File: app/src/main/java/com/example/stas/oxforddictionary/presentation/view/entry/EntryFragment.java
package com.example.stas.oxforddictionary.presentation.view.entry;


import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;
import com.example.stas.oxforddictionary.R;
import com.example.stas.oxforddictionary.presentation.adapter.DefinitionAdapter;
import com.example.stas.oxforddictionary.presentation.adapter.Item;
import com.example.stas.oxforddictionary.presentation.presenter.EntryPresenter;
import com.example.stas.oxforddictionary.presentation.view.main.IMainActivity;
import java.io.IOException;
import java.util.List;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class EntryFragment extends Fragment implements EntryContract.View {

    @BindView(R.id.wordEntryET) EditText wordEntry;
    @BindView(R.id.searchSubmitBtn) Button searchBtn;
    @BindView(R.id.definition_recycler_view) RecyclerView definitionRecyclerView;
    @BindView(R.id.wordInfoContainer) LinearLayout infoContainer;
    @BindView(R.id.titleText) TextView word;
    @BindView(R.id.titleTranscription) TextView transcription;
    @BindView(R.id.titleSound) ImageButton soundBtn;
    @BindView(R.id.entryProgressBar) ProgressBar progressBar;
    @BindView(R.id.synonymsBtn) Button synonymsBtn;
    @BindView(R.id.examplesBtn) Button examplesBtn;
    private Unbinder unbinder;
    private EntryContract.Presenter presenter;
    private RecyclerView.LayoutManager layoutManager;
    private DefinitionAdapter definitionAdapter;
    private Animation moveUp;
    private IMainActivity mainActivity;

    public EntryFragment() {

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_entry, container, false);
        unbinder = ButterKnife.bind(this, view);
        presenter = new EntryPresenter(this);
        moveUp = AnimationUtils.loadAnimation(getActivity(),
                R.anim.move_up);

        searchBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (wordEntry.length() > 0){
                    presenter.getDefinition(wordEntry.getText().toString());
                    infoContainer.setVisibility(View.INVISIBLE);
                    progressBar.setVisibility(View.VISIBLE);
                }else {
                    Toast.makeText(getContext(), "word is missing", Toast.LENGTH_SHORT).show();
                }
            }
        });
        soundBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                progressBar.setVisibility(View.VISIBLE);
                presenter.getSound(word.getText().toString());
            }
        });

        synonymsBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
            String wordId = word.getText().toString();
            mainActivity.navigateToSynonyms(getContext(), wordId);
            }
        });

        examplesBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });

        layoutManager = new LinearLayoutManager(getContext());
        definitionAdapter = new DefinitionAdapter();
        definitionRecyclerView.setLayoutManager(layoutManager);
        definitionRecyclerView.setAdapter(definitionAdapter);
        return view;
    }

    @Override
    public void showDefinition(List<Item> definitions, List<String> titleSet) {
        synonymsBtn.setVisibility(View.VISIBLE);
        examplesBtn.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.GONE);
        word.setText(titleSet.get(0));
        transcription.setText(titleSet.get(1));
        infoContainer.setVisibility(View.VISIBLE);
        infoContainer.startAnimation(moveUp);
        definitionAdapter.setItems(definitions);
        definitionAdapter.notifyDataSetChanged();
    }

    @Override
    public void playSound(String soundURL) {
        MediaPlayer mediaPlayer = new MediaPlayer();
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        try {
            mediaPlayer.setDataSource(soundURL);
        } catch (IOException e) {
            e.printStackTrace();
        }
        try {
            mediaPlayer.prepare();
        } catch (IOException e) {
            //todo handle error
            e.printStackTrace();
        }
        mediaPlayer.start();
    }

    @Override
    public void hideProgressBar() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showError(String errorMessage) {
        Snackbar.make(infoContainer, errorMessage, Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mainActivity = (IMainActivity) getActivity();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
        presenter.detachView();
    }
}

File: app/src/main/java/com/example/stas/oxforddictionary/presentation/view/entry/EntryFragment.kt
package com.example.stas.oxforddictionary.presentation.view.entry

import android.content.Context
import android.media.MediaPlayer
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.*
import com.example.stas.oxforddictionary.App
import com.example.stas.oxforddictionary.R
import com.example.stas.oxforddictionary.data.database.model.ViewedWordModel
import com.example.stas.oxforddictionary.presentation.view.entry.adapter.AutocompleteAdapter
import com.example.stas.oxforddictionary.presentation.view.entry.adapter.DefinitionAdapter
import com.example.stas.oxforddictionary.presentation.view.entry.adapter.Item
import com.example.stas.oxforddictionary.presentation.view.main.IMainActivity
import com.example.stas.oxforddictionary.presentation.view.main.MainActivity
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.fragment_entry.*
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.function.Function
import javax.inject.Inject

class EntryFragment : Fragment(), EntryContract.View {

    @Inject
    lateinit var presenter: EntryContract.Presenter
    @Inject
    lateinit var definitionAdapter: DefinitionAdapter
    @Inject
    lateinit var autocompleteAdapter: AutocompleteAdapter
    private var moveUp: Animation? = null
    private var mainActivity: IMainActivity? = null
    lateinit var subject: PublishSubject<String>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_entry, container, false)
        (activity!!.application as App).createEntryComponent().injectEntryFragment(this)
        presenter.attachView(this)
        moveUp = AnimationUtils.loadAnimation(activity,
                R.anim.move_up)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        subject = PublishSubject.create<String>()

        wordEntryET.addTextChangedListener(object: TextWatcher{
            override fun afterTextChanged(s: Editable?) {
                subject.debounce(1000, TimeUnit.MILLISECONDS)
                        .switchMap { presenter.onEntryTextChanged(wordEntryET.text.toString()) }
                        .map { mapper: List<ViewedWordModel> -> mapper.map { it.word } }
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe{t-> showAutocompletes(t)}
               subject.onNext(wordEntryET.text.toString())
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }
        })

        searchSubmitBtn.setOnClickListener {
            autocompleteAdapter.clear()
            autocompleteAdapter.notifyDataSetChanged()
            (activity as MainActivity).closeKeyboard()

            if (wordEntryET.length() > 0) {
                presenter.getDefinition(wordEntryET.text.toString())
                wordInfoContainer.visibility = View.INVISIBLE
                progressBar.visibility = View.VISIBLE
            } else {
                Toast.makeText(context, "word is missing", Toast.LENGTH_SHORT).show()
            }

        }
        soundImBtn.setOnClickListener {
            progressBar.visibility = View.VISIBLE
            presenter.getSound(title.text.toString())
        }

        synonymsBtn.setOnClickListener {
            val wordId = title.text.toString()
            mainActivity!!.navigateToSynonyms(context, wordId)
        }

        examplesBtn.setOnClickListener {
            val wordId = title.text.toString()
            mainActivity!!.navigateToExamples(context, wordId)
        }
        autocompleteRec.layoutManager = LinearLayoutManager(context)
        autocompleteRec.adapter = autocompleteAdapter


        definitionRW.layoutManager = LinearLayoutManager(context)
        definitionAdapter.starClickListener = { definition ->
            presenter.saveDefinition(title.text.toString(), definition)
            Toast.makeText(context, "$definition добавлен", Toast.LENGTH_SHORT).show()
        }
        definitionRW.adapter = definitionAdapter
    }

    override fun showDefinition(definitions: List<Item>, titleSet: List<String?>) {
        synonymsBtn.visibility = View.VISIBLE
        examplesBtn.visibility = View.VISIBLE
        progressBar.visibility = View.GONE
        title.text = titleSet[0]
        transcription.text = titleSet[1]
        wordInfoContainer.visibility = View.VISIBLE
        wordInfoContainer.startAnimation(moveUp)
        definitionAdapter.setItems(definitions)
        definitionAdapter.notifyDataSetChanged()
    }

    override fun playSound(soundURL: String?) {
        val mediaPlayer = MediaPlayer()
        try {
            mediaPlayer.setDataSource(soundURL)
        } catch (e: IOException) {
            showError(getString(R.string.set_source_error))
        }
        try {
            mediaPlayer.prepare()
        } catch (e: IOException) {
            showError(getString(R.string.player_error))
        }
        mediaPlayer.start()
    }

    override fun showAutocompletes(t: List<String>?) {
        t?.let { autocompleteAdapter.setList(it)
        autocompleteAdapter.notifyDataSetChanged()}
    }

    override fun hideProgressBar() {
        progressBar.visibility = View.GONE
    }

    override fun showError(errorMessage: String) {
        Snackbar.make(wordInfoContainer, errorMessage, Snackbar.LENGTH_SHORT).show()
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mainActivity = activity as IMainActivity?
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.detachView()
    }
}
